### PR TITLE
[do not merge] custom derive for record types

### PR DESCRIPTION
**Pull request by dicej**
This is a first draft of an implementation of
https://github.com/bytecodealliance/wasmtime/issues/4308, i.e. derive macros for
`ComponentType`, `Lift`, and `Lower` for composite types in the component model.

It borrows heavily from the work @jameysharp did in
https://github.com/bytecodealliance/wasmtime/pull/4217.  Thanks, Jamey!

Thanks also to @fibonacci1729 for pairing up on this.

This draft only covers records; I expect the other composite types will follow a
similar pattern.  I'm looking to get feedback on the overall approach before I
continue.

Known (potential) issues:

- Currently, the implementations for `ComponentType`, `Lift`, and `Lower`
  duplicate a fair amount of code, so I plan to address that.

- I've made `ComponentTypesBuilder::record_type` public for testing purposes.
  Let me know if that's a problem or if there's a better way.

- Also for testing: I've made `Memory::wasmtime_memory` and `Store::inner`
  crate-public.

- I've added a `test` module to crates/wasmtime/src/component/func/typed.rs for
  testing the derives.  Let me know if there's a better place for it.

Signed-off-by: Joel Dice <joel.dice@fermyon.com>

### CODE REVIEW

**Comment by dicej** [comment]
@alexcrichton @jameysharp @fibonacci1729 This is my first draft of an implementation of https://github.com/bytecodealliance/wasmtime/issues/4308, supporting just `record`s so far.  Please let me me know what you think of the overall approach.

File: crates/component-macro/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+
+    static UNEXPECTED_STYLE_MESSAGE: &str =
+        "unrecognized component type keyword (expected `record`, `variant`, `enum`, or `union`)";
+
+    static UNEXPECTED_SYNTAX_MESSAGE: &str = "expected `(<style>)` syntax";
+
+    let mut style = None;
+
+    for attribute in &input.attrs {
+        if attribute.path.leading_colon.is_none() && attribute.path.segments.len() == 1 {
```
Comment (alexcrichton) [inline crates/component-macro/src/lib.rs:28]: Personally for patterns like this I find it convenient to have "early return" to avoid nested indentation getting too unwieldy. For example:

```rust
if attribute.path.leading_colon.is_some() || attribute.path.segments.len() != 1 {
    continue;
}

// ... everything else minus one indentation level
```

File: crates/component-macro/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,382 @@
+use proc_macro2::{Literal, TokenStream, TokenTree};
+use quote::quote;
+use syn::{parse_macro_input, parse_quote, Data, DeriveInput, Error, Result};
+
+static INCOMPATIBLE_TYPE_MESSAGE: &str =
+    "a `record` must be defined as a `struct` with at least one named field";
```
Comment (alexcrichton) [inline crates/component-macro/src/lib.rs:6]: Personally I think it's ok to inline all these at the use-site, otherwise I find myself tabbing back and forth to what exactly this error message is saying

File: crates/component-macro/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+                types.extend(quote!(#ty,));
+
+                let index = syn::Index::from(index);
+
+                field_initializations.extend(quote!(#ident: result.#index,));
+            }
+
+            delegate_record_lift(
+                input,
+                quote!((#types)),
```
Comment (alexcrichton) [inline crates/component-macro/src/lib.rs:181]: Oh this is clever! I hadn't actually thought of simply delegating to the tuple implementation. That definitely makes this easier since you don't have to do offset calculations here in the derive macro.

That being said though the main limitation of this is that it won't work for structs with > 16 fields. That doesn't seem too common though so otherwise seems like a good way to start?

File: crates/component-macro/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    } else {
+        return Err(Error::new(name.span(), INCOMPATIBLE_TYPE_MESSAGE));
+    };
+
+    match &body.fields {
+        syn::Fields::Named(fields) => {
+            let values = fields
+                .named
+                .iter()
+                .map(|syn::Field { ident, .. }| quote!(self.#ident.clone(),))
```
Comment (alexcrichton) [inline crates/component-macro/src/lib.rs:256]: Instead of cloning could this pass `&T` since `&T` is `Lower` if `T` is `Lower`?

File: crates/component-macro/src/lib.rs (hunk 3 above)
Comment (alexcrichton) [reply crates/component-macro/src/lib.rs:181]: One thing you may want to explore though is what the error messages look like if a field of the record doesn't implement the requisite trait. For example what's the compiler error message when you derive `Lift` for a record with a `String` in it?

**Review by alexcrichton — COMMENTED** [review]
Seems like a reasonable start to me! I think it'd be good to test these externally ratehr than internally but otherwise this seems like a good start 👍 

I like the idea of delegating to tuples but that may not "scale well" in the long-term due to the 16-field limitation and perhaps additionally due to compiler error messages, but for now getting off the ground with them seems reasonable.

File: crates/component-macro/src/lib.rs (hunk 3 above)
Comment (dicej) [reply crates/component-macro/src/lib.rs:181]: The cleverness is due to @jameysharp -- that's straight from [his patch](https://github.com/bytecodealliance/wasmtime/compare/main...jameysharp:component-derive).

Comment (jameysharp) [reply crates/component-macro/src/lib.rs:177]: I'm certainly enjoying being called clever! But before I got sidetracked, I had intended to drop this trick. :laughing:

In #4217 I factored out `next_field` to make both tuple and record macro expansions relatively straightforward. While `lower` and `lift` were already pretty easy, `next_field` wraps up the common math for `size`, `store`, and `load`. So I hope it should be simple now to almost just copy `impl_component_ty_for_tuples`, except as a proc-macro instead of `macro_rules`, and with field names instead of indexes.

As I recall, this bit of cleverness is also what forced me to call `.clone()` in `lower`: I couldn't build a tuple of borrows, because the implementation I delegated to expected the borrow to be on the outside of the tuple instead. Skipping the indirection should let you remove those clones.

File: crates/component-macro/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+            let mut lowers = TokenStream::new();
+            let mut stores = TokenStream::new();
+
+            for syn::Field { ty, ident, .. } in &fields.named {
+                lowers.extend(quote!(wasmtime::component::Lower::lower(
+                    &self.#ident, store, options, unsafe {
+                        // Extract a mutable pointer to the field and transmute it to a `&mut MaybeUninit<#ty>`,
+                        // taking care to never create a `&mut Self::Lower` or a `&mut #ty`, which would be
+                        // unsound.
+                        std::mem::transmute::<_, _>(std::ptr::addr_of_mut!((*dst.as_mut_ptr()).#ident))
```
Comment (alexcrichton) [inline crates/component-macro/src/lib.rs:275]: I'm not actually sure if this is possible, but if we can I think it'd be best to reuse the `map_maybe_uninit!` macro used internally here as well. If that doesn't work out though can you fill in the two type parameters to `transmute` here to clarify what's happening?

**Review by alexcrichton — APPROVED** [review]
This is looking fantastic to me! I'd be happy to merge with a few more tests namely around negative things and some renaming, but otherwise looks great!

For the `const _: () = ..` I think what you can do is basically:

```rust
fn expand_the_trait() -> Result<TokenStream> {
    let prev = /* all the previous logic you already have */;
    Ok(quote!(const _: () = { #prev };))
}
```

and that should do the trick. The idea is that new definitions are all placed into an "anonymous" namespace so the outer crate doesn't have access to them (which we don't want)

File: crates/component-macro/src/lib.rs (hunk 5 above)
Comment (dicej) [reply crates/component-macro/src/lib.rs:275]: Yeah, I started out reusing `map_maybe_uninit!`, but that required adding a `#[macro_export]` and making `MaybeUninitExt` public.  If that sounds okay to you, I'll do that.

Comment (alexcrichton) [reply crates/component-macro/src/lib.rs:275]: That should be fine as long as they're all marked `#[doc(hidden)]` to not actually show up in the public documentation

File: crates/component-macro/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+                    .unwrap_or_else(|| Literal::string(&ident.as_ref().unwrap().to_string()));
+
+                field_names_and_checks.extend(
+                    quote!((#literal, <#ty as wasmtime::component::ComponentType>::typecheck),),
+                );
+
+                sizes.extend(quote!(#internal::next_field::<#ty>(&mut size);));
+
+                alignments.extend(
+                    quote!(align = align.max(<#ty as wasmtime::component::ComponentType>::align());)
```
Comment (jameysharp) [inline crates/component-macro/src/lib.rs:379]: Random thought: for alignment, you could de-duplicate the field types, e.g. a struct with a bunch of `u32`s has the same alignment as one `u32`. I think it might marginally improve compile times overall. It's certainly not necessary to do though.

Overall, this looks good to me! It's up to Alex on final review though.

**Comment by dicej** [comment]
I've restructured the code a bit to reduce duplication.  Please let me know if there's anything else to change (e.g. more test cases) before this would be considered merge-able.

**Review by jameysharp — APPROVED** [review]
My opinion is that this is good to merge as it stands.

I speculate that it might work out better in the end to drive `expand` according to whether the input is a struct or an enum, rather than whether the expected output is a record or whatever. And I think you could move more of the common logic from the outer `proc_macro_derive` functions into `expand`.

But in my opinion, I think it's worth landing this now and continuing on. I expect we'll get a better idea of how to minimize duplication once you've implemented more of the component interface types.

**Comment by dicej** [comment]
Sorry for all the commit churn.  I've added the test Jamey suggested, rebased onto main, and squashed everything down to a single commit.

**Comment by jameysharp** [comment]
Regarding the current CI failures: I think you need to add `#[doc(hidden)]` to `impl StoreOpaque` in `crates/wasmtime/src/store.rs` since that type is now technically `pub` through the `__internal` mod.
